cmd/object-api-utils.go

```diff
@@ -675,3 +675,60 @@ func CleanMinioInternalMetadataKeys(metadata map[string]string) map[string]strin
 	}
 	return newMeta
 }
+
+// snappyCompressReader compresses data as it's read
+// from the underlying io.Reader.
+type snappyCompressReader struct {
+	r      io.Reader
+	w      *snappy.Writer
+	closed bool
+	buf    bytes.Buffer
+}
+
+func newSnappyCompressReader(r io.Reader) *snappyCompressReader {
+	cr := &snappyCompressReader{r: r}
+	cr.w = snappy.NewBufferedWriter(&cr.buf)
+	return cr
+}
+
+func (cr *snappyCompressReader) Read(p []byte) (int, error) {
+	if cr.closed {
+		// if snappy writer is closed r has been completely read,
+		// return any remaining data in buf.
+		return cr.buf.Read(p)
+	}
+
+	for {
```

Comment:
@vcabbage i cannot understand why we need a for loop here. I see all the io operations already does this for us. It already does the Read([]byte) iteratively.

For eg, If you see https://golang.org/src/io/io.go?s=12796:12856#L353 the copyBuffer already reads iteratively.

Is there any other reason that i cannot see here?

Reply:
In this way, we are overriding the default behavior of an IO call. There would be just one exhaustive Read([]byte) call which reads fully until the actual reader runs out of data.

Reply:
I could have sworn this fixed an issue when running the test cases, but I can't reproduce it so I'm going to chalk it up to me misinterpreting the failure.